File: day11_robot/python/day11_robot_part_1.py
#!/bin/python

import sys
import io
from computer import IntCodeComputer

class Robot:
    def __init__(self, init_mem):
        self.computer = IntCodeComputer(init_mem, [0])
        self.pos = (0, 0)
        self.direction = (0, -1)
        self.painted_panels = []
    
    def get_panel_colour(self):
        for (coords, colour) in self.painted_panels:
            if coords == self.pos:
                return colour
        return 0 # black
    
    def paint_panel(self, colour):
        self.painted_panels = \
            [(coords, colour) for (coords, colour) in self.painted_panels if coords != self.pos] + \
            [(self.pos, colour)]

    def run(self):
        self.computer.inputs = [self.get_panel_colour()]
        self.computer.run()
        if self.computer.is_halted():
            return False
        else:
            [new_colour, new_direction] = self.computer.outputs
            self.paint_panel(new_colour)
            (xd, yd) = self.direction
            if new_direction == 0:
                # turn left
                self.direction = (-yd, xd)
            elif new_direction == 1:
                # turn right
                self.direction = (yd, -xd)
            else:
                raise Exception("Unrecognised direction code")
            (x, y) = self.pos
            (xd, yd) = self.direction
            self.pos = (x + xd, y + yd)
            return True


file_path = "../data/puzzle_input.csv"
init_mem = []
with io.open(file_path, "r") as f:
    line = f.readline()
    if line:
        init_mem = [int(x.strip()) for x in line.split(",")]
    else:
        raise Exception("No program in file!")

robot = Robot(init_mem)
while robot.run():
    pass


print("Output: {}\n".format(len(robot.painted_panels)))

File: day11_robot/erlang/day11_robot_part_1.erl
-module(day11_robot_part_1).

-import(computer, [load_program/1, set_inputs/2, exec/1, get_outputs/1, clear_outputs/1, get_state/1]).
-export([paint_hull/0]).


-record(robot, {
    x,
    y,
    xd,
    yd
}).


paint_hull() ->
    {ok, Device} = file:open("../data/puzzle_input.csv", [read]),
    Computer = try computer:load_program(Device)
      after file:close(Device)
    end,
    map_size(do_paint_hull(Computer, #robot{x=0, y=0, xd=0, yd=-1}, #{})).


get_panel_colour(Hull, X, Y) ->
    try maps:get({X, Y}, Hull)
    catch 
      error:_ -> 0
    end.


paint_panel(Hull, X, Y, Colour) ->
    maps:put({X, Y}, Colour, Hull).


do_paint_hull(Computer, #robot{x=X, y=Y, xd=XD, yd=YD}, Hull) ->
    Colour = get_panel_colour(Hull, X, Y),
    NewComputer = computer:exec(computer:set_inputs(Computer, [Colour])),
    case computer:get_state(NewComputer) of
        halted -> 
            Hull;
        _ ->
            [NewColour, Turn] = computer:get_outputs(NewComputer),
            NewHull = paint_panel(Hull, X, Y, NewColour),
            NewRobot = case Turn of
                0 -> #robot{x=X + YD, y=Y - XD, xd=YD, yd=-XD};
                1 -> #robot{x=X - YD, y=Y + XD, xd=-YD, yd=XD}
            end,
            do_paint_hull(computer:clear_outputs(NewComputer), NewRobot, NewHull)
    end.